### PR TITLE
Adding custom provision provider

### PR DESCRIPTION
**Pull request by Petermarcu**
Submitting this as a conversation starter. The goal is to enable people to not use any of the built in providers but instead allow them to execute the provisioning tools they want to use. This basically makes a dummy "custom" provider that no-ops the main steps but pre and post hook commands still run. This enables people to do what they need at the right stages and not be blocked because they don't want to use bicep or terraform. 

@weikanglim @ellismg @jongio @rajeshkamal5050

### CODE REVIEW

**Comment by vhvb1989** [comment]
This is interesting.  I wonder if we really need a `custom` provider for enabling `hooks`.
I think folks can use `pre-deploy` or `post-package` instead (from a project w/o infra).  

I wonder if it would make more sense to have an `ephemeral` or `void` provider  for this.  IMO, `custom` sounds like the customer can define how it would behave, but it reality, this is a `no-op` provider

**Comment by Petermarcu** [comment]
I was also wondering if we should just let someone launch a process here.

I think what you described mostly does what I was looking for but I'm not sure about the contract of emitting a empty arm template. I literally don't want the deployment to happen via the ARM call in this case.

**Comment by weikanglim** [comment]
What is azd's role in provisioning, now that provisioning itself is delegated away? Would running `<my provision tool>` achieve the same outcome? Is azd a facade?

**Comment by Petermarcu** [comment]
Its always been intended to be partially a facade in the sense that we want to enable people to do whatever they need to do while still providing a consistent interface for tools above it. We also intend to make provisioning in different ways better and better by investing in concrete implementations of them like bicep and TF where the tool can provide more insight in places where there is first class integration. The key to me is that the first class integration can't get in the way of people doing something else as part of the flow.

**Comment by jongio** [comment]
As a developer, I would like to be able to:

1. Bypass any azd command
2. Provide a custom implementation for any azd command


**Bypass:** 
 - azd augments azure.yaml schema to allow the user to signal the command is bypassed

**Custom:**
- azd allows the user to define a hook for commands:


for example:

```
hooks:
  - provision:
```

If the hook is defined then that is run instead of the azd command.

This type of design would allow us to implement this behavior for any command, not just provision.

Aside: A less desirable UX (IMO) would be, instruct the user to have a hook/command no-op script if they want to bypass a command.

**Comment by wbreza** [comment]
> ```
> hooks:
>   - provision:
> ```
> 
> If the hook is defined then that is run instead of the azd command.
> 
> This type of design would allow us to implement this behavior for any command, not just provision.

Yes! This is inline with what I was thinking as well.  Today the hooks only expose `pre` and `post`.  If a user wants to take control over any project/service command they should simply be able to set that in the hooks configuration and take control of that process.  

This same concept can then be easily applied to any other commands we want to enable this level of flexibility.

```yaml
hooks:
  # Run this script in place of default azd implementation
  provision: 
    run: ./scripts/provision.sh
```

File: cli/azd/pkg/infra/provisioning/custom/custom.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// Package custom contains a custom implementation of provider.Provider that no-ops the provisioning behavior
+// allowing people to use the hooks to implement custom behavior. This provider is registered for use when
+// this package is imported, and can be imported for side effects only to register the provider, e.g.:
+package custom
```
Comment (wbreza) [inline cli/azd/pkg/infra/provisioning/custom/custom.go:7]: I agree with @vhvb1989 here.  We shouldn't **need** to implement a custom provider for a noop / override scenario.  My thinking is more inline with @jongio 's approach to extending this behavior globally in hook configuration.

**Comment by wbreza** [comment]
And in the case of a real `noop` use case our default provision provider (bicep) maybe shouldn't error if it doesn't find what it needs.  If a `main.bicep` or other required files doesn't exist azd could respond with something like:

`SKIPPED - No IaC Templates found` 

With a warning message to add IaC templates into the `infra` folder and try again.  Returning a noop here would still allow an `azd up` experience to work without having to create a custom provider or hook noop implementation.

**Comment by savannahostrowski** [comment]
In my mind, this is really very similar to what we had planned for provision extensibility in some ways. Would also be interesting to think about how something like this fit in with other plans we had in this space (e.g. Pulumi). In my mind, both require us "delegating" provisioning to something else, getting outputs and then using them to deploy etc.

**Comment by Petermarcu** [comment]
Agreed. I think in some cases we would delegate and expect a contract with the component. In these cases, we would know more and could have a more feature rich experience.

So are we concluding that we should do what Jon suggested? Allow a hook for all the mainline commands that lets you run a command instead of the standard action?

For the skip case. what do people think? Should it just be an empty hook for provision?

**Comment by savannahostrowski** [comment]
I think extending hooks to surface core azd commands seems like a really nice UX. I also like @wbreza's suggestion for a noop case. The main thing I want to make sure of is that no additional overhead/YAML is needed in azure.yaml if you just want to do the "vanilla" azd thing and default to our command implementations/definitions.  The UX should also remain unchanged in that case.

Extra YAML should only be required if you want to customize/override the experience.

**Comment by Petermarcu** [comment]
Yup, that makes sense. I'm a bit sad that bicep is the default and that the "easier" solutions were going to implement in the future will require more yaml :(

**Comment by savannahostrowski** [comment]
I think that with some of the abstraction work we are thinking about the Bicep bits could be hidden from a developer on Day 0, FWIW.

**Comment by Petermarcu** [comment]
> I think that with some of the abstraction work we are thinking about the Bicep bits could be hidden from a developer on Day 0, FWIW.

Yeah, I think we're likely going to have to make the "default" provision provider more than just Bicep to keep the yaml clean for the simple case.

**Comment by Petermarcu** [comment]
In fact, I'm wondering if we should start forcing people to put "bicep" in their azure.yaml right away. Start with a deprecation warning and then move it to a requirement at some point?

**Comment by savannahostrowski** [comment]
I think that's probably going to happen as we start to identify the exact UX we want for abstraction.

**Comment by Petermarcu** [comment]
So more questions here:

How does someone specify "noop"?
    Is it by overriding with the hook and specifying nothing?
    Should they set the string to "noop"?
    Should they be able to do it without hooks and specify noop as the infra type?

If someone specifies "noop" however we decide that is done:

Should the pre and post hooks for that command still run?

**Comment by vhvb1989** [comment]
Here's my proposal.
See the current behavior and how I would change it to support no-op in an easy way (and backwards compatible):

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:Calibri;font-size:11.0pt'>


<div style='direction:ltr'>



azure.yaml - infra field | Current behavior | Change to:
-- | -- | --
Empty - | Expectation:         folder: infra     file: infra/main.bicep             Use bicep provider    ** Fail if    expectations are missing | Smart detection:       Use folder infra if present.    Do no-op if infra folder not present. Hooks        for `pre-provision` and `post-provision` can still be set and are dispatched.                 When infra   folder is present:       Look for either (using        priority order):         main.bicep     terraform     cdk / other        Fallback to no-op if infra folder is empty or nothing        known was found.                  ** azd never fails. It always fall   back to no-op as worst case.
Infra.module =   <module>   Infra.path =   <path>   Infra.provider =   <provider> | Expectation:         folder: <path>     file: infra/<module>.[bicep/tf]     Provider: <provider>             Use  <provider>    ** Fail if    expectations are missing | No smart detection.       Azd follows the explicit         user-definitions.     Use infra.provider         = null to use         the no-op provider.             ** azd fails if something is missing
Combination of:       Infra.module = [empty \| module]   Infra.path = [empty \| path]   Infra.provider = [empty \| provider] | Use a default for each empty field | Do smart-detection when only <path> is        set:         try to detect the provider         and module or fall-back to no-op.        Do explicit-expectation when <module> and/or        <provider> are set.         Fail if missing config.



</div>


</body>

</html>

**Comment by savannahostrowski** [comment]
I think that the behaviour of an empty infra field in azure.yaml may need to change as we think about abstraction. Empty in a world where we allow a user to potentially define only services (with hosts, languages, paths; manually or via the simplified init) could in theory leverage smart defaults + Bicep modules "generated" out into tmp/ for provision. This would result in no infra/ at all in the project root, not visible to the user.

Not saying that's the decided solution but arguably an option since we are also already generating IaC in the new simplified init flow and persisting it to the project.

**Comment by Petermarcu** [comment]
In that case, would they set an alternate location for the infra folder where all the generated stuff will be?

**Comment by savannahostrowski** [comment]
Not sure if you're asking Victor or me :) 

I'll elaborate anyway... 

My initial thinking is for a Day 0 experience, where someone is coming off of the simplified init flow, there is no infra/ folder that the user has to think about. They just have to think about azure.yaml and azd handles the rest. 

I'm imagining that if a user wanted to expand/eject out of the abstracted, day 0 experience because they now want to tweak configurations and want to take ownership of the IaC bits, they could run something like `azd infra synth`  to spit out the Bicep files we'd been using "under the hood" and land them in an experience very similar to what it looks like today (but there's also composability in the picture so management of resources gets a bit easier). The directory we choose to expand the IaC out into could be infra/ by default but overridable via a prompt under that command.

IMO, this overriding of provision provider/no-op scenario is only really applicable to folks who have "expanded"/"ejected". In that case, I think we'd probably want something like @vhvb1989 has in the top-right box. I think that addresses both abstracted and non-abstracted views of IaC.

**Comment by vhvb1989** [comment]
@savannahostrowski 

> they could run something like azd infra synth to spit out the Bicep files we'd been using "under the hood"

If `azd` spits out the IaC to `/infra` since day-0, we don't need to have this extra `azd infra synth`. 

Also, azd would use either ARM/bicep under the hood anyway.  I don't see why to keep it hidden.  

Having an azd-only abstraction for IaC can't be scale-up to all Azure w/o a high cost.  What would it be its limits?  
Is it a few popular services for getting-started samples ?

**Comment by savannahostrowski** [comment]
The piece to consider here is that these IaC files are noise and overwhelming to folks that don't understand Azure/Bicep/IaC. They just want to go from A->B.

**Comment by Petermarcu** [comment]
I agree that we want to move in a direction where, if the info for the infra is provided in some other form and is generated, those become intermediate build artifacts and not put into the source tree.

**Comment by ellismg** [comment]
I'm going to close this for now - the pressing issues for this specific feature have been addressed in other ways.  We're still thinking about extensibility in this area, but there's no clear step forward for this PR right now.  Happy to have the discussion continue here, however.

**Pull request closed without merging**